src/components/common/DatePickerWithRange/index.tsx

```diff
@@ -0,0 +1,97 @@
+"use client";
+
+import { format } from "date-fns";
+import { Calendar as CalendarIcon } from "lucide-react";
+import * as React from "react";
+import { DateRange } from "react-day-picker";
+
+import { Button } from "~/components/ui/button";
+import { Calendar } from "~/components/ui/calendar";
+import {
+  Popover,
+  PopoverContent,
+  PopoverTrigger,
+} from "~/components/ui/popover";
+import { cn } from "~/lib/utils";
+
+interface DataItem {
+  id: number;
+  name: string;
+  date: {
+    from: Date;
+    to: Date;
+  };
+}
+
+interface DataProperties {
+  className?: React.HTMLAttributes<HTMLDivElement>;
+  data: DataItem[];
+}
+
+export function DatePickerWithRange({ className, data }: DataProperties) {
+  const [date, setDate] = React.useState<DateRange | undefined>();
+
+  function filterDateRanges(
+    data: DataItem[],
+    selectedDate: DateRange,
+  ): DataItem[] {
+    const filteredData = data.filter(
+      (item) =>
+        selectedDate?.from &&
+        item.date.from.getTime() >= selectedDate?.from.getTime() &&
+        selectedDate?.to &&
+        item.date.to.getTime() <= selectedDate?.to.getTime(),
+    );
+    return filteredData;
```

Comment:
```suggestion
    return data.filter(
      (item) =>
        selectedDate?.from &&
        item.date.from.getTime() >= selectedDate?.from.getTime() &&
        selectedDate?.to &&
        item.date.to.getTime() <= selectedDate?.to.getTime(),
    );
```

Reply:
**selectedDate.from' is possibly 'undefined'.ts(18048)**

I added the 'selectedDate.from' check because of the error above

Reply:
fixed